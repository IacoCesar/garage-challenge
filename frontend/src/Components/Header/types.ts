export interface IHeader {
  testID?: string
  isHomeScreen?: boolean
  title?: string
  callHandleBack?: (navigation: any) => void
  callHandleChangeVisualMode?: (themeSelector: string) => void
}

export interface IProps extends IHeader {
  themeSelector: string
}
