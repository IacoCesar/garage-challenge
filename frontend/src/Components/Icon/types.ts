export interface IProps {
  testID?: string
  size: number
  name: string
  color: string
  type?: string
}
