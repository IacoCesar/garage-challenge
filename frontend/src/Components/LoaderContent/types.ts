export interface IProps {
  testID?: string
  size: number
  color: string
}
