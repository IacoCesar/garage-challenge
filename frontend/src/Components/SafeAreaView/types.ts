export interface IProps {
  testID?: string
  children: JSX.Element
}

export interface IThemeProps {
  themeSelector: string
}
