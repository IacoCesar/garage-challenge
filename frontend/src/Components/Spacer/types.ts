export interface IProps {
  testID?: string
  min: number
  max: number
}
