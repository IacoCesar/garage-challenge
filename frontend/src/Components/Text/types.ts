export interface ITextProps {
  color?: string
  fontSize?: number
  isBold?: boolean
}

export interface IProps extends ITextProps {
  label: string
  testID?: string
}
