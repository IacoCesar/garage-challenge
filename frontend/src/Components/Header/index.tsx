import React from 'react'
import S from './styles'
import Text from '@/Components/Text'
import { IHeader } from './types'
import Icon from '@/Components/Icon'
import { theme } from '@/Helpers/Theme'
import { useSelector } from 'react-redux'
import { RootState } from '@/Store'
import { assignTestId } from '@/Helpers/TestId'

const Component: React.FC<IHeader> = ({
  testID = 'Header',
  isHomeScreen = false,
  title = '',
  callHandleBack,
  callHandleChangeVisualMode
}) => {
  const themeSelector = useSelector((state: RootState) => state.themeData.theme)

  return (
    <S.Container
      {...assignTestId('View', `${testID}_Container`)}
      themeSelector={themeSelector}
    >
      <S.Content {...assignTestId('View', `${testID}_Container_Content`)}>
        {!isHomeScreen && (
          <S.ButtonLeftItem
            {...assignTestId(
              'TouchableOpacity',
              `${testID}_Container_Content_ButtonLeftItem`
            )}
            onPress={callHandleBack}
          >
            <Icon
              testID={`${testID}_Container_Content_ButtonLeftItem_Icon`}
              size={20}
              name={'arrow-back-ios'}
              type="material"
              color={theme.standard[themeSelector].icon}
            />
          </S.ButtonLeftItem>
        )}
        <S.ContentTitle
          {...assignTestId('View', `${testID}_Container_Content_ContentTitle`)}
          isHomeScreen={isHomeScreen}
          themeSelector={themeSelector}
        >
          <Text
            testID={`${testID}_Container_Content_ContentTitle_Text`}
            fontSize={25}
            isBold
            label={isHomeScreen ? 'Garage' : title}
          />
        </S.ContentTitle>
      </S.Content>
      {isHomeScreen && callHandleChangeVisualMode && (
        <S.ButtonLeftItem
          {...assignTestId(
            'TouchableOpacity',
            `${testID}_Container_ButtonLeftItem`
          )}
          onPress={() => callHandleChangeVisualMode(themeSelector)}
        >
          <Icon
            testID={`${testID}_Container_ButtonLeftItem_Icon`}
            size={20}
            name={'brightness-2'}
            type="material"
            color={theme.standard[themeSelector].textColor}
          />
        </S.ButtonLeftItem>
      )}
    </S.Container>
  )
}

export default Component
