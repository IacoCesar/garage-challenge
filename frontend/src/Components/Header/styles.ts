import styled from 'styled-components/native'
import { IProps } from './types'

const Container = styled.View<IProps>`
  display: flex;
  height: 72px;
  border: 0px solid transparent;
  padding-horizontal: ${({ theme }) => theme.spacing.global.horizontal};
  background-color: ${({ theme, themeSelector }) =>
    theme.standard[themeSelector].background};
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`

const Content = styled.View`
  flex: 1;
  flex-direction: row;
  align-items: center;
`

const ContentTitle = styled.View<IProps>`
  flex: 1;
  justify-content: center;
  align-items: flex-start;
  shadow-opacity: 0.5;
  shadow-radius: 5px;
  shadow-color: ${({ theme, themeSelector }) =>
    theme.standard[themeSelector].shadow};
  shadow-offset: 0px 0px;
`

const ButtonLeftItem = styled.TouchableOpacity`
  display: flex;
  width: 45px;
  height: 45px;
  justify-content: center;
  align-items: center;
`

const ButtonRightItem = styled.TouchableOpacity`
  display: flex;
  width: 45px;
  height: 45px;
  justify-content: center;
  align-items: center;
`

export default {
  Container,
  Content,
  ContentTitle,
  ButtonLeftItem,
  ButtonRightItem
}
