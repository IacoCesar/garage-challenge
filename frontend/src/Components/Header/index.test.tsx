import React from 'react'
import { IHeader } from './types'
import { mockTestId } from '@/Helpers/TestId'
import Header from '.'
import { render, fireEvent } from '@/Helpers/TestHelpers'
import { act } from 'react-test-renderer'

describe('Header', () => {
  let props: IHeader = {
    testID: 'Header',
    title: 'Test',
    isHomeScreen: false,
    callHandleBack: jest.fn(),
    callHandleChangeVisualMode: jest.fn()
  }

  const containerElement = mockTestId(
    'View',
    `${props.testID!}_Container`
  ).testID
  const buttonBackElement = mockTestId(
    'TouchableOpacity',
    `${props.testID!}_Container_Content_ButtonLeftItem`
  ).testID
  const titleContentElement = mockTestId(
    'View',
    `${props.testID!}_Container_Content_ContentTitle`
  ).testID

  const buttonThemeElement = mockTestId(
    'TouchableOpacity',
    `${props.testID!}_Container_ButtonLeftItem`
  ).testID

  it(`should render ${containerElement}`, async () => {
    const wrapper = render(<Header {...props} />)
    const el_container = wrapper.getByTestId(containerElement)

    expect(el_container).toBeTruthy()
    expect(el_container.children.length).toBe(1)
  })

  it(`should find title content ${titleContentElement}`, async () => {
    const wrapper = render(<Header {...props} />)
    const title = await wrapper.findByTestId(titleContentElement)

    expect(title).toBeTruthy()
  })

  it(`should onPress ${buttonBackElement}`, async () => {
    const wrapper = render(<Header {...props} />)
    const button = await wrapper.findByTestId(buttonBackElement)

    await act(async () => {
      fireEvent.press(button)
      expect(props.callHandleBack).toHaveBeenCalled()
    })
  })

  it(`should onPress ${buttonThemeElement}`, async () => {
    const wrapper = render(<Header {...props} isHomeScreen={true} />)

    const button = await wrapper.findByTestId(buttonThemeElement)

    await act(async () => {
      fireEvent.press(button)
      expect(props.callHandleBack).toHaveBeenCalled()
    })
  })

  it('should render snapshot', async () => {
    const wrapper = render(<Header {...props} />)

    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
