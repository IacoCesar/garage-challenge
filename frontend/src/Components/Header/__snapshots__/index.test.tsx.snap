// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render snapshot 1`] = `
<View
  accessible={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFF",
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderWidth": 0,
        "display": "flex",
        "flexDirection": "row",
        "height": 72,
        "justifyContent": "space-between",
        "paddingHorizontal": 16,
      },
    ]
  }
  testID="View_Header_Container"
  themeSelector="default"
>
  <View
    accessible={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
    testID="View_Header_Container_Content"
  >
    <View
      accessible={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "height": 45,
          "justifyContent": "center",
          "opacity": 1,
          "width": 45,
        }
      }
      testID="TouchableOpacity_Header_Container_Content_ButtonLeftItem"
    >
      <View
        accessible={false}
        color="#000000"
        name="arrow-back-ios"
        size={20}
        testID="View_Header_Container_Content_ButtonLeftItem_Icon"
      />
    </View>
    <View
      accessible={false}
      isHomeScreen={false}
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.5,
            "shadowRadius": 5,
          },
        ]
      }
      testID="View_Header_Container_Content_ContentTitle"
      themeSelector="default"
    >
      <View
        accessible={false}
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
        testID="View_Header_Container_Content_ContentTitle_Text"
      >
        <Text
          accessible={false}
          color="#000000"
          fontSize={25}
          isBold={true}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 25,
                "fontWeight": "bold",
              },
            ]
          }
          testID="Text_Header_Container_Content_ContentTitle_Text_Content_Text"
        >
          Test
        </Text>
      </View>
    </View>
  </View>
</View>
`;
